builtin/logical/pki/periodic.go

```diff
@@ -0,0 +1,176 @@
+package pki
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/helper/consts"
+)
+
+const (
+	minUnifiedTransferDelay = 30 * time.Minute
+)
+
+type unifiedTransferStatus struct {
+	isRunning  atomic.Bool
+	lastRun    time.Time
+	forceRerun atomic.Bool
+}
+
+func (uts *unifiedTransferStatus) forceRun() {
+	uts.forceRerun.Store(true)
+}
+
+func newUnifiedTransferStatus() *unifiedTransferStatus {
+	return &unifiedTransferStatus{}
+}
+
+// runUnifiedTransfer meant to run as a background, this will process all and
+// send all missing local revocation entries to the unified space if the feature
+// is enabled.
+func runUnifiedTransfer(sc *storageContext) {
+	b := sc.Backend
+	status := b.unifiedTransferStatus
+
+	isPerfStandby := b.System().ReplicationState().HasState(consts.ReplicationDRSecondary | consts.ReplicationPerformanceStandby)
+
+	if isPerfStandby || b.System().LocalMount() {
+		// We only do this on active enterprise nodes, when we aren't a local mount
+		return
+	}
+
+	config, err := b.crlBuilder.getConfigWithUpdate(sc)
+	if err != nil {
+		b.Logger().Error("failed to retrieve crl config from storage for unified transfer background process",
+			"error", err)
+		return
+	}
+
+	if !status.lastRun.IsZero() {
+		// We have run before, we only run again if we have
+		// been requested to forceRerun, and we haven't run since our
+		// minimum delay
+		if !(status.forceRerun.Load() && time.Since(status.lastRun) < minUnifiedTransferDelay) {
+			return
+		}
+	}
+
+	if !config.UnifiedCRL {
+		// Feature is disabled, no need to run
+		return
+	}
+
+	clusterId, err := b.System().ClusterID(sc.Context)
+	if err != nil {
+		b.Logger().Error("failed to fetch cluster id for unified transfer background process",
+			"error", err)
+		return
+	}
+
+	if !status.isRunning.CompareAndSwap(false, true) {
+		b.Logger().Debug("an existing unified transfer process is already running")
+		return
+	}
+	defer status.isRunning.Store(false)
+
+	// Reset our flag before we begin, we do this before we start as
```

Comment:
Not sure I follow here? Wouldn't we want to set this to `forceRerun.Store(true)` to force a re-issue, if we don't explicitly set `forceRerun.Store(false)` later, when `err == nil`? 

My reading is that, with the above (`status.forceRerun.Load() && ...`), this would do the opposite of what you want.

Reply:
Maybe I need to rework the comment then but I believe this is right. What I'm attempting to protect against is the `forceRun()` from `revokeCert` call being swallowed incorrectly at the end of this function if it is called halfway through execution.

If we set it to false when `err == nil` we will have a race, in which this loop finished successfully, but the revokeCert got an error, flags us and we then overwrite with the false based on our `err == nil`

Reply:
Ah yeah, I see now. I think its fine.